File: frontend/src/services/chordRecognition.ts
/**
 * Enhanced Chord Recognition Engine
 * 
 * This module provides comprehensive chord recognition and parsing capabilities
 * supporting various chord formats and notations.
 */

export interface ChordComponents {
  root: string;              // Root note (C, D, E, etc.)
  accidental?: string;       // Sharp (#) or flat (b)
  quality?: string;          // Major (empty), minor (m), diminished (dim), augmented (aug)
  extension?: string;        // 7, 9, 11, 13, maj7, etc.
  suspension?: string;       // sus2, sus4
  addition?: string;         // add9, add11, etc.
  modification?: string;     // b5, #5, b9, #9, etc.
  bassNote?: string;         // Slash chord bass note (C/E -> bassNote: "E")
}

export interface ParsedChord {
  original: string;          // Original chord notation
  normalized: string;        // Normalized chord notation
  components: ChordComponents;
  isValid: boolean;
  quality: 'major' | 'minor' | 'diminished' | 'augmented' | 'suspended' | 'unknown';
  enharmonicEquivalents: string[];
  errors?: string[];
}

/**
 * Enhanced chord recognition patterns
 */
const CHORD_PATTERNS = {
  // Root note patterns
  root: /^([A-G])/,
  
  // Accidental patterns (sharp or flat)
  accidental: /^[A-G]([#b])/,
  
  // Quality patterns (more comprehensive)
  quality: /^[A-G][#b]?(m|min|major|maj|dim|aug|°|o|\+)/i,
  
  // Extension patterns (comprehensive for jazz and complex chords)
  extension: /(?:maj|min|dim|aug|M)?([67]|9|11|13|add[29]|add11|add13)/i,
  
  // Suspension patterns
  suspension: /(sus[24]?)/i,
  
  // Modification patterns (altered dominants, etc.)
  modification: /([#b][59]|[#b]11|[#b]13)/g,
  
  // Slash chord pattern
  slashChord: /\/([A-G][#b]?)/,
  
  // Full comprehensive pattern for validation (more restrictive)
  full: /^[A-G][#b]?(?:(?:maj|min|dim|aug|sus[24]?|add[29]|add1[13]|M|m)?(?:[67]|9|11|13)?(?:[#b][59]|[#b]1[13])*(?:\/[A-G][#b]?)?)?$/
};

/**
 * Enharmonic equivalents mapping
 */
const ENHARMONIC_EQUIVALENTS: Record<string, string[]> = {
  'C#': ['Db'],
  'Db': ['C#'],
  'D#': ['Eb'],
  'Eb': ['D#'],
  'F#': ['Gb'],
  'Gb': ['F#'],
  'G#': ['Ab'],
  'Ab': ['G#'],
  'A#': ['Bb'],
  'Bb': ['A#']
};

/**
 * Note names in different languages/notations
 */
const NOTE_TRANSLATIONS = {
  // Spanish/Latin notation
  'Do': 'C',
  'Re': 'D',
  'Mi': 'E',
  'Fa': 'F',
  'Sol': 'G',
  'La': 'A',
  'Si': 'B',
  // German notation
  'H': 'B'
};

/**
 * Quality detection patterns
 */
const QUALITY_PATTERNS = {
  major: /^[A-G][#b]?(?!.*(?:m|min|dim|aug|°|o|\+|sus))/i,
  minor: /^[A-G][#b]?(?:m|min)(?!aj)/i,
  diminished: /^[A-G][#b]?(?:dim|°|o)(?![a-z])/i,
  augmented: /^[A-G][#b]?(?:aug|\+)/i,
  suspended: /^[A-G][#b]?.*sus/i
};

/**
 * Enhanced chord recognition class
 */
export class ChordRecognitionEngine {
  /**
   * Parse a chord notation into its components
   */
  public parseChord(input: string): ParsedChord {
    const original = input.trim();
    const errors: string[] = [];
    
    if (!original) {
      return this.createInvalidChord(original, ['Empty chord notation']);
    }

    // Normalize input (handle different notations)
    let normalized = this.normalizeChordNotation(original);
    
    // Basic validation using comprehensive pattern
    if (!CHORD_PATTERNS.full.test(normalized)) {
      errors.push(`Invalid chord format: ${original}`);
      return this.createInvalidChord(original, errors);
    }

    // Extract components
    const components = this.extractComponents(normalized);
    
    // Determine chord quality
    const quality = this.determineQuality(normalized);
    
    // Get enharmonic equivalents
    const enharmonicEquivalents = this.getEnharmonicEquivalents(components.root + (components.accidental || ''));
    
    return {
      original,
      normalized,
      components,
      isValid: true,
      quality,
      enharmonicEquivalents
    };
  }

  /**
   * Validate if a chord notation is valid
   */
  public isValidChord(chord: string): boolean {
    return this.parseChord(chord).isValid;
  }

  /**
   * Get all possible enharmonic equivalents for a chord
   */
  public getEnharmonicEquivalents(note: string): string[] {
    // Handle both uppercase and preserve proper case for flats/sharps
    const normalizedNote = note.charAt(0).toUpperCase() + note.slice(1);
    return ENHARMONIC_EQUIVALENTS[normalizedNote] || [];
  }

  /**
   * Normalize chord notation (handle different languages/formats)
   */
  private normalizeChordNotation(chord: string): string {
    let normalized = chord.trim();
    
    // Handle Spanish/Latin notation
    Object.entries(NOTE_TRANSLATIONS).forEach(([foreign, english]) => {
      const pattern = new RegExp(`^${foreign}(?![a-z])`, 'i');
      normalized = normalized.replace(pattern, english);
    });
    
    // Normalize quality indicators
    normalized = normalized
      .replace(/major/gi, 'maj')
      .replace(/minor/gi, 'min')
      .replace(/°/g, 'dim')
      .replace(/\+/g, 'aug');
    
    return normalized;
  }

  /**
   * Extract chord components
   */
  private extractComponents(chord: string): ChordComponents {
    const components: ChordComponents = {
      root: ''
    };

    // Extract root note
    const rootMatch = chord.match(CHORD_PATTERNS.root);
    if (rootMatch) {
      components.root = rootMatch[1];
    }

    // Extract accidental
    const accidentalMatch = chord.match(CHORD_PATTERNS.accidental);
    if (accidentalMatch) {
      components.accidental = accidentalMatch[1];
    }

    // Extract quality
    const qualityMatch = chord.match(CHORD_PATTERNS.quality);
    if (qualityMatch) {
      components.quality = qualityMatch[1];
    }

    // Extract extension
    const extensionMatch = chord.match(CHORD_PATTERNS.extension);
    if (extensionMatch) {
      components.extension = extensionMatch[1];
    }

    // Extract suspension
    const suspensionMatch = chord.match(CHORD_PATTERNS.suspension);
    if (suspensionMatch) {
      components.suspension = suspensionMatch[1];
    }

    // Extract modifications
    const modificationMatches = chord.match(CHORD_PATTERNS.modification);
    if (modificationMatches) {
      components.modification = modificationMatches.join(',');
    }

    // Extract bass note (slash chord)
    const slashMatch = chord.match(CHORD_PATTERNS.slashChord);
    if (slashMatch) {
      components.bassNote = slashMatch[1];
    }

    return components;
  }

  /**
   * Determine chord quality
   */
  private determineQuality(chord: string): ParsedChord['quality'] {
    if (QUALITY_PATTERNS.suspended.test(chord)) return 'suspended';
    if (QUALITY_PATTERNS.diminished.test(chord)) return 'diminished';
    if (QUALITY_PATTERNS.augmented.test(chord)) return 'augmented';
    if (QUALITY_PATTERNS.minor.test(chord)) return 'minor';
    if (QUALITY_PATTERNS.major.test(chord)) return 'major';
    return 'unknown';
  }

  /**
   * Create invalid chord result
   */
  private createInvalidChord(original: string, errors: string[]): ParsedChord {
    return {
      original,
      normalized: original,
      components: { root: '' },
      isValid: false,
      quality: 'unknown',
      enharmonicEquivalents: [],
      errors
    };
  }

  /**
   * Batch parse multiple chords
   */
  public parseChords(chords: string[]): ParsedChord[] {
    return chords.map(chord => this.parseChord(chord));
  }

  /**
   * Extract chords from ChordPro content
   */
  public extractChordsFromContent(content: string): ParsedChord[] {
    const chordPattern = /\[([^\]]+)\]/g;
    const chords: string[] = [];
    let match;
    
    while ((match = chordPattern.exec(content)) !== null) {
      const chordName = match[1].trim();
      if (chordName && !chords.includes(chordName)) {
        chords.push(chordName);
      }
    }
    
    return this.parseChords(chords);
  }

  /**
   * Validate ChordPro content and return detailed analysis
   */
  public validateChordProContent(content: string): {
    isValid: boolean;
    totalChords: number;
    validChords: number;
    invalidChords: ParsedChord[];
    uniqueRoots: string[];
    qualities: Record<string, number>;
  } {
    const parsedChords = this.extractChordsFromContent(content);
    const invalidChords = parsedChords.filter(chord => !chord.isValid);
    const validChords = parsedChords.filter(chord => chord.isValid);
    
    // Analyze unique roots
    const uniqueRoots = [...new Set(validChords.map(chord => 
      chord.components.root + (chord.components.accidental || '')
    ))];
    
    // Analyze qualities
    const qualities: Record<string, number> = {};
    validChords.forEach(chord => {
      qualities[chord.quality] = (qualities[chord.quality] || 0) + 1;
    });
    
    return {
      isValid: invalidChords.length === 0,
      totalChords: parsedChords.length,
      validChords: validChords.length,
      invalidChords,
      uniqueRoots,
      qualities
    };
  }
}

// Export singleton instance
export const chordRecognitionEngine = new ChordRecognitionEngine();

// Legacy compatibility exports
export function isValidChord(chord: string): boolean {
  return chordRecognitionEngine.isValidChord(chord);
}

export function parseChord(chord: string): ParsedChord {
  return chordRecognitionEngine.parseChord(chord);
}

File: backend/chordme/chord_recognition.py
"""
Enhanced Chord Recognition Engine for Python Backend

This module provides comprehensive chord recognition and parsing capabilities
supporting various chord formats and notations.
"""

import re
from typing import Dict, List, Optional, Tuple, Any
from dataclasses import dataclass


@dataclass
class ChordComponents:
    """Structure representing chord components"""
    root: str = ""
    accidental: Optional[str] = None
    quality: Optional[str] = None
    extension: Optional[str] = None
    suspension: Optional[str] = None
    addition: Optional[str] = None
    modification: Optional[str] = None
    bass_note: Optional[str] = None


@dataclass
class ParsedChord:
    """Structure representing a parsed chord"""
    original: str
    normalized: str
    components: ChordComponents
    is_valid: bool
    quality: str  # 'major', 'minor', 'diminished', 'augmented', 'suspended', 'unknown'
    enharmonic_equivalents: List[str]
    errors: Optional[List[str]] = None


class ChordRecognitionEngine:
    """Enhanced chord recognition engine for Python backend"""
    
    # Enhanced chord patterns
    CHORD_PATTERNS = {
        'root': re.compile(r'^([A-G])'),
        'accidental': re.compile(r'^[A-G]([#b])'),
        'quality': re.compile(r'^[A-G][#b]?(maj|min|dim|aug|m|°|o|\+)(?![a-z])', re.IGNORECASE),
        'extension': re.compile(r'(?:maj|min|dim|aug|M)?([67]|9|11|13|add[29]|add11|add13)', re.IGNORECASE),
        'suspension': re.compile(r'(sus[24]?)', re.IGNORECASE),
        'modification': re.compile(r'([#b][59]|[#b]11|[#b]13)'),
        'slash_chord': re.compile(r'/([A-G][#b]?)'),
        'full': re.compile(r'^[A-G][#b]?(?:(?:maj|min|major|dim|aug|sus[24]?|add[29]|add1[13]|M|m|°|o|\+)?(?:[67]|9|11|13)?(?:[#b][59]|[#b]1[13])*(?:/[A-G][#b]?)?)?$', re.IGNORECASE)
    }
    
    # Enharmonic equivalents mapping
    ENHARMONIC_EQUIVALENTS = {
        'C#': ['Db'],
        'Db': ['C#'],
        'D#': ['Eb'],
        'Eb': ['D#'],
        'F#': ['Gb'],
        'Gb': ['F#'],
        'G#': ['Ab'],
        'Ab': ['G#'],
        'A#': ['Bb'],
        'Bb': ['A#']
    }
    
    # Note translations for different languages/notations
    NOTE_TRANSLATIONS = {
        # Spanish/Latin notation
        'Do': 'C',
        'Re': 'D',
        'Mi': 'E',
        'Fa': 'F',
        'Sol': 'G',
        'La': 'A',
        'Si': 'B',
        # German notation
        'H': 'B'
    }
    
    # Quality detection patterns
    QUALITY_PATTERNS = {
        'major': re.compile(r'^[A-G][#b]?(?!.*(?:m|min|dim|aug|°|o|\+))', re.IGNORECASE),
        'minor': re.compile(r'^[A-G][#b]?(?:m|min)(?!aj)', re.IGNORECASE),
        'diminished': re.compile(r'^[A-G][#b]?(?:dim|°|o)', re.IGNORECASE),
        'augmented': re.compile(r'^[A-G][#b]?(?:aug|\+)', re.IGNORECASE),
        'suspended': re.compile(r'^[A-G][#b]?.*sus', re.IGNORECASE)
    }
    
    def parse_chord(self, input_chord: str) -> ParsedChord:
        """Parse a chord notation into its components"""
        original = input_chord.strip()
        errors = []
        
        if not original:
            return self._create_invalid_chord(original, ['Empty chord notation'])
        
        # Normalize input (handle different notations)
        normalized = self._normalize_chord_notation(original)
        
        # Basic validation using comprehensive pattern
        if not self.CHORD_PATTERNS['full'].match(normalized):
            errors.append(f'Invalid chord format: {original}')
            return self._create_invalid_chord(original, errors)
        
        # Extract components
        components = self._extract_components(normalized)
        
        # Determine chord quality
        quality = self._determine_quality(normalized)
        
        # Get enharmonic equivalents
        root_with_accidental = components.root + (components.accidental or '')
        enharmonic_equivalents = self.get_enharmonic_equivalents(root_with_accidental)
        
        return ParsedChord(
            original=original,
            normalized=normalized,
            components=components,
            is_valid=True,
            quality=quality,
            enharmonic_equivalents=enharmonic_equivalents
        )
    
    def is_valid_chord(self, chord: str) -> bool:
        """Validate if a chord notation is valid"""
        return self.parse_chord(chord).is_valid
    
    def get_enharmonic_equivalents(self, note: str) -> List[str]:
        """Get all possible enharmonic equivalents for a note"""
        # Handle both uppercase and preserve proper case for flats/sharps
        normalized_note = note[0].upper() + note[1:] if len(note) > 1 else note.upper()
        return self.ENHARMONIC_EQUIVALENTS.get(normalized_note, [])
    
    def _normalize_chord_notation(self, chord: str) -> str:
        """Normalize chord notation (handle different languages/formats)"""
        normalized = chord.strip()
        
        # Handle Spanish/Latin notation
        for foreign, english in self.NOTE_TRANSLATIONS.items():
            pattern = re.compile(f'^{foreign}(?![a-z])', re.IGNORECASE)
            normalized = pattern.sub(english, normalized)
        
        # Normalize quality indicators
        normalized = re.sub(r'major', 'maj', normalized, flags=re.IGNORECASE)
        normalized = re.sub(r'minor', 'min', normalized, flags=re.IGNORECASE)
        normalized = re.sub(r'dim', 'dim', normalized, flags=re.IGNORECASE)
        normalized = re.sub(r'aug', 'aug', normalized, flags=re.IGNORECASE)
        normalized = normalized.replace('°', 'dim')
        normalized = normalized.replace('o', 'dim')
        normalized = normalized.replace('+', 'aug')
        
        return normalized
    
    def _extract_components(self, chord: str) -> ChordComponents:
        """Extract chord components"""
        components = ChordComponents()
        
        # Extract root note
        root_match = self.CHORD_PATTERNS['root'].match(chord)
        if root_match:
            components.root = root_match.group(1)
        
        # Extract accidental
        accidental_match = self.CHORD_PATTERNS['accidental'].match(chord)
        if accidental_match:
            components.accidental = accidental_match.group(1)
        
        # Extract quality
        quality_match = self.CHORD_PATTERNS['quality'].match(chord)
        if quality_match:
            components.quality = quality_match.group(1)
        
        # Extract extension
        extension_match = self.CHORD_PATTERNS['extension'].search(chord)
        if extension_match:
            components.extension = extension_match.group(1)
        
        # Extract suspension
        suspension_match = self.CHORD_PATTERNS['suspension'].search(chord)
        if suspension_match:
            components.suspension = suspension_match.group(1)
        
        # Extract modifications
        modification_matches = self.CHORD_PATTERNS['modification'].findall(chord)
        if modification_matches:
            components.modification = ','.join(modification_matches)
        
        # Extract bass note (slash chord)
        slash_match = self.CHORD_PATTERNS['slash_chord'].search(chord)
        if slash_match:
            components.bass_note = slash_match.group(1)
        
        return components
    
    def _determine_quality(self, chord: str) -> str:
        """Determine chord quality"""
        if self.QUALITY_PATTERNS['suspended'].match(chord):
            return 'suspended'
        if self.QUALITY_PATTERNS['diminished'].match(chord):
            return 'diminished'
        if self.QUALITY_PATTERNS['augmented'].match(chord):
            return 'augmented'
        if self.QUALITY_PATTERNS['minor'].match(chord):
            return 'minor'
        if self.QUALITY_PATTERNS['major'].match(chord):
            return 'major'
        return 'unknown'
    
    def _create_invalid_chord(self, original: str, errors: List[str]) -> ParsedChord:
        """Create invalid chord result"""
        return ParsedChord(
            original=original,
            normalized=original,
            components=ChordComponents(),
            is_valid=False,
            quality='unknown',
            enharmonic_equivalents=[],
            errors=errors
        )
    
    def parse_chords(self, chords: List[str]) -> List[ParsedChord]:
        """Batch parse multiple chords"""
        return [self.parse_chord(chord) for chord in chords]
    
    def extract_chords_from_content(self, content: str) -> List[ParsedChord]:
        """Extract chords from ChordPro content"""
        chord_pattern = re.compile(r'\[([^\]]+)\]')
        chords = []
        unique_chords = set()
        
        for match in chord_pattern.finditer(content):
            chord_name = match.group(1).strip()
            if chord_name and chord_name not in unique_chords:
                chords.append(chord_name)
                unique_chords.add(chord_name)
        
        return self.parse_chords(chords)
    
    def validate_chordpro_content(self, content: str) -> Dict[str, Any]:
        """Validate ChordPro content and return detailed analysis"""
        parsed_chords = self.extract_chords_from_content(content)
        invalid_chords = [chord for chord in parsed_chords if not chord.is_valid]
        valid_chords = [chord for chord in parsed_chords if chord.is_valid]
        
        # Analyze unique roots
        unique_roots = list(set(
            chord.components.root + (chord.components.accidental or '')
            for chord in valid_chords
            if chord.components.root
        ))
        
        # Analyze qualities
        qualities = {}
        for chord in valid_chords:
            quality = chord.quality
            qualities[quality] = qualities.get(quality, 0) + 1
        
        return {
            'is_valid': len(invalid_chords) == 0,
            'total_chords': len(parsed_chords),
            'valid_chords': len(valid_chords),
            'invalid_chords': [
                {
                    'chord': chord.original,
                    'errors': chord.errors or []
                }
                for chord in invalid_chords
            ],
            'unique_roots': unique_roots,
            'qualities': qualities
        }


# Create singleton instance
chord_recognition_engine = ChordRecognitionEngine()


# Legacy compatibility functions
def is_valid_chord(chord: str) -> bool:
    """Legacy compatibility function for chord validation"""
    return chord_recognition_engine.is_valid_chord(chord)


def parse_chord(chord: str) -> Dict[str, Any]:
    """Legacy compatibility function for chord parsing"""
    result = chord_recognition_engine.parse_chord(chord)
    return {
        'original': result.original,
        'normalized': result.normalized,
        'is_valid': result.is_valid,
        'quality': result.quality,
        'components': {
            'root': result.components.root,
            'accidental': result.components.accidental,
            'quality': result.components.quality,
            'extension': result.components.extension,
            'suspension': result.components.suspension,
            'bass_note': result.components.bass_note
        },
        'enharmonic_equivalents': result.enharmonic_equivalents,
        'errors': result.errors
    }